### ISSUE TITLE
Overlay immediately closes with rootClose in shadow root

### ISSUE THREAD

**Issue by angrycat9000** [react-restart/ui#90]
## Describe the bug

`Overlay` with `rootClose` closes as soon as it is opened if it is hosted inside of a [custom element](https://developer.mozilla.org/en-US/docs/Web/API/Web_Components/Using_custom_elements) shadow root.

I'm using  custom elements to isolate some react based components that are hosted in an older application.  The `Overlay` immediately closes when opened if it is hosted inside of a shadow root of a custom element.  The same code works correctly when hosted in the light dom.

## To Reproduce

1. Create a custom element with a shadow root that hosts a react render root
2. Render a Overlay with `rootClose` and `onHide` set.  Also have the `container` option point to an element in the shadow root
3. Click the button to open the overlay.

## Reproducible Example

https://codesandbox.io/s/bold-dan-5kfh5r - Declares a custom element in `index.js` and has it render the same React component as is rendered in the light DOM.  The custom element embedded version does not work, the light DOM version does.

- Click the _Show in shadow root_ button.  Note the popover flashes and then disappears.
- Click _Show in light DOM_ button.  Note that the popover appears and remains visibile.

## Expected behavior

The popover remains open until an additional click.

## Screenshots

[Screencast from 08-08-2023 10:04:38 AM.webm](https://github.com/react-restart/ui/assets/48362463/5d0aacd5-e550-4091-942c-e513579b0af7)

## Environment (please complete the following information)

- Operating System: Linux
- Browser, Version: Chrome 115.0.5790.110
![image](https://github.com/react-restart/ui/assets/48362463/0c7e5740-fbab-4747-a109-dccb7a2cc93b)

**Comment by kyletsang** [comment]
The original workaround that @jquense added to fix this doesn't work in the shadow DOM because `currentEvent` is undefined

https://github.com/react-restart/ui/blob/44dc346744b37c33bd0d0309e5cd827751fce9c4/src/useClickOutside.ts#L104C9-L104C9

As a result, the click handler is processing twice

**Comment by jquense** [comment]
Funny I just saw that material UI has a different work around for the same issue in their ClickAway component, maybe we can try that?

**Comment by kyletsang** [comment]
> Funny I just saw that material UI has a different work around for the same issue in their ClickAway component, maybe we can try that?

Unfortunately couldn't get tests to pass with the MUI approach.  I'll need to drop this for now, since I'm short on time ATM.

**Comment by jquense** [comment]
fair enough, i don't think this is a super pressing use case anyway!

**Comment by mklekowski** [comment]
Hi, any update? Do you need some help to make test passing? I'd like to make it work as soon as possible as the whole react-bootstrap is not working in the shadowDOM.

**Comment by angrycat9000** [comment]
FYI @mklekowski  I ended up working around this by not using `rootClose` and then adding my own event listeners for Escape and a click outside the overlay.

**Comment by mklekowski** [comment]
Actually, I have an issue with DropdownButton, when clicking two events are triggered, first for opening and second for closing. Probably it's also possible to disable rootClose and implement it manually, but I prefer to fix it in the library directly.

@jquense can you look at the issue?

**Comment by mklekowski** [comment]
Also, the code is using [Window/event](https://developer.mozilla.org/en-US/docs/Web/API/Window/event) which is deprecated. Maybe it can be implemented in different way?

**Comment by HighLiuk** [comment]
> Actually, I have an issue with DropdownButton, when clicking two events are triggered, first for opening and second for closing. Probably it's also possible to disable rootClose and implement it manually, but I prefer to fix it in the library directly.

@mklekowski do you have a workaround? I have the same issue as you, with `DropdownButton`.

**Comment by mklekowski** [comment]
Yes and now. The workaround for me was to change usage of the library to [mui/base-ui](https://mui.com/base-ui/getting-started/) - it has better support for [shadow dom](https://mui.com/material-ui/customization/shadow-dom/) and it looks like it's better maintained.

**Comment by HighLiuk** [comment]
> Yes and no :) The workaround for me was to change usage of the library to [mui/base-ui](https://mui.com/base-ui/getting-started/) - it has better support for [shadow dom](https://mui.com/material-ui/customization/shadow-dom/) and it looks like it's better maintained.

Ouch - thank you anyway - I'm bound to Bootstrap though, so hopefully this gets fixed soon. My current workaround is

```tsx
    <Dropdown
      autoClose="outside"
      show={showDropdown}
      onToggle={(show, meta) => {
        const target = meta.originalEvent?.target as HTMLElement;

        if (target.tagName !== "DROP-DOWN") { // "drop-down" is my custom element name
          setShowDropdown(show);
        }
      }}
    >
    </Dropdown>
```

This makes the Dropdown kinda work, but it doesn't close the menu when you click _inside_ the custom element but _outside_ the button.
